### DISCUSSION TITLE
Memory leaks

### DISCUSSION THREAD

**Discussion by MaximilianWilke** [IIC2133-PUC/2024-2#25] [Tarea 0] [answered]
Hola trabajando con el input PART-2-easy-2 y compare mi output con el correcto y vi que robaba menos puffles del que tenia que robar, entonces hice unos prints y me di cuenta que los colores de algunos puffles (quiza el 1%) aparecian con valores al azares (cosa que no ocurria con los inputs de la parte 1 y Part-2-easy-1) por lo que creo que debe ser un problema de memoria, entonces trate de usar valgrind pero no se bien como interpretarlo (no se si se puede subir un imagen del output de valgrind al github)

**Comment by tomastrivino** [comment]
Hola, sí, por favor sube una imagen del output de valgrind para ayudarte. No incluyas código de tu tarea.

**Comment by MaximilianWilke** [comment]
<img width="638" alt="image" src="https://github.com/user-attachments/assets/f3e56f37-bc18-4e55-9a24-93c56fc00337">

**Comment by tomastrivino** [comment, marked as answer]
Efectivamente tu programa tiene problemas de memoria, tal como se aprecia en el output de Valgrind (fíjate en `ERROR SUMMARY` y en el error `Segmentation fault`, justo abajo del primero). Es bastante posible que estos problemas sean los causantes de los outputs aleatorios que mencionas, por lo que es esencial que soluciones aquellos leaks antes de probar otras opciones.

El consejo que te podría dar es que revises muy bien cuándo estás haciendo `free()`. **Desde el momento en que liberas una dirección de memoria, estás seguro que no será utilizada nuevamente dentro de tu programa** (por ejemplo, si hago `free(pinguino_1)`, no debería intentar acceder a `pinguino_1` nuevamente). Revisa bien en qué partes estás liberando memoria. Este tipo de problemas también puede ocurrir si tienes un array e intentas acceder a un índice que está "fuera del rango". 

¡Éxito!

**Comment by MaximilianWilke** [comment]
Pero no comprendo porque esto sucede únicamente este input.txt. Si probara con los otros no debiera ocurrir lo mismo de los colores?

**Reply by tomastrivino** [reply]
Podría tratarse de un caso borde. Es decir, que justo para ese input se caiga en alguna parte.

**Comment by MaximilianWilke** [comment]
Hola, sigo teniendo el mismo problema, revisé con el input más basico y Valgrind dice lo mismo (aunque sin el cambio de colores de los puffles),  lo raro es que la linea 34 de main es un scanf para leer la capacidad de la tienda (y es la linea que valgrind me dice que revise), y eso un codigo que venia hecho. En la ayudantía de hoy me dijeron que podía tener relación con como se comportaba el codigo dependiendo de si no todos los pinguinos se iban de la tienda al finalizar la lectura de inputs, pero mirando pareciera que el comportamiento sí funciona

**Reply by tomastrivino** [reply]
Hola,
por favor sube tu código para probarlo. Lo que te puedo decir es que hay un problema con el código base, ya que definimos inicialmente `char color`. Puedes cambiarlo a `char *color` o `char color[NRO_DE_CARACTERES]` (recomendado) para ver si eso arregla algo.

Saludos

**Comment by MaximilianWilke** [comment]
listo lo subi

**Reply by tomastrivino** [reply]
Probé tu código
<img width="842" alt="image" src="https://github.com/user-attachments/assets/74f747b4-4863-4272-88ae-a55f2cfc72ba">

El problema que me sale a mí es este, el que indica que no liberaste la memoria.

**Reply by MaximilianWilke** [reply]
a mi no me sale eso, como lo corriste?

**Reply by tomastrivino** [reply]
```
valgrind --leak-check=full ./puffle src/puffle/input.txt out.txt
```
ese src/puffle/input.txt es el quee tienes tú en la carpeta (no hace falta subir los tests a github btw)

**Comment by MaximilianWilke** [comment]
Logre eliminar el error en tienda.c: 29 pero el main.c: 78 persiste, alguna idea?

<img width="544" alt="image" src="https://github.com/user-attachments/assets/746e5ad1-4d09-4a1a-9c47-aa70d7f6da9b">
